### ISSUE TITLE
Is InP model for "Inpainting" - Because if so...

### ISSUE THREAD

**Issue by AILova** [aigc-apps/VideoX-Fun#203]
I'm curious if the InP model is for in painting or something else.  I aske because every video I've seen on is showing how it supports First and Last frame reference images, which to me, is quite different than inpainting.  I'm trying to see if the Wan Fun model, for example, will in context in paint areas in a video masked with some inpaint mask color (e.g., in VACE is the gray color that gets inpainted in the video...).  So my question is, does the Wan Fun 1.3b and/or 14B models do true video inpainting and if so, can someone point to an example workflow in ComfyUI?  Thanks in advance.

**Comment by AILova** [comment]
Exactly, this is what I'm trying to figure out.  If not, why is it called InPaint?

**Comment by loboere** [comment]
I think they call it that because it does an InPaint from the first frame to the last. It's like in Vace putting gray from the first frame to the last, but in Fun, you can't InPaint certain areas.

**Comment by bubbliiiing** [comment]
Our code uses the Inpainting approach during training. In fact, we can set up masks similar to square blocks for video editing. This is our mask generation strategy.
```
def get_random_mask(shape, image_start_only=False):
    f, c, h, w = shape
    mask = torch.zeros((f, 1, h, w), dtype=torch.uint8)

    if not image_start_only:
        if f != 1:
            mask_index = np.random.choice([0, 1, 2, 3, 4, 5, 6, 7, 8, 9], p=[0.05, 0.2, 0.2, 0.2, 0.05, 0.05, 0.05, 0.1, 0.05, 0.05]) 
        else:
            mask_index = np.random.choice([0, 1], p = [0.2, 0.8])
        if mask_index == 0:
            center_x = torch.randint(0, w, (1,)).item()
            center_y = torch.randint(0, h, (1,)).item()
            block_size_x = torch.randint(w // 4, w // 4 * 3, (1,)).item()  # 方块的宽度范围
            block_size_y = torch.randint(h // 4, h // 4 * 3, (1,)).item()  # 方块的高度范围

            start_x = max(center_x - block_size_x // 2, 0)
            end_x = min(center_x + block_size_x // 2, w)
            start_y = max(center_y - block_size_y // 2, 0)
            end_y = min(center_y + block_size_y // 2, h)
            mask[:, :, start_y:end_y, start_x:end_x] = 1
        elif mask_index == 1:
            mask[:, :, :, :] = 1
        elif mask_index == 2:
            mask_frame_index = np.random.randint(1, 5)
            mask[mask_frame_index:, :, :, :] = 1
        elif mask_index == 3:
            mask_frame_index = np.random.randint(1, 5)
            mask[mask_frame_index:-mask_frame_index, :, :, :] = 1
        elif mask_index == 4:
            center_x = torch.randint(0, w, (1,)).item()
            center_y = torch.randint(0, h, (1,)).item()
            block_size_x = torch.randint(w // 4, w // 4 * 3, (1,)).item()  # 方块的宽度范围
            block_size_y = torch.randint(h // 4, h // 4 * 3, (1,)).item()  # 方块的高度范围

            start_x = max(center_x - block_size_x // 2, 0)
            end_x = min(center_x + block_size_x // 2, w)
            start_y = max(center_y - block_size_y // 2, 0)
            end_y = min(center_y + block_size_y // 2, h)

            mask_frame_before = np.random.randint(0, f // 2)
            mask_frame_after = np.random.randint(f // 2, f)
            mask[mask_frame_before:mask_frame_after, :, start_y:end_y, start_x:end_x] = 1
        elif mask_index == 5:
            mask = torch.randint(0, 2, (f, 1, h, w), dtype=torch.uint8)
        elif mask_index == 6:
            num_frames_to_mask = random.randint(1, max(f // 2, 1))
            frames_to_mask = random.sample(range(f), num_frames_to_mask)

            for i in frames_to_mask:
                block_height = random.randint(1, h // 4)
                block_width = random.randint(1, w // 4)
                top_left_y = random.randint(0, h - block_height)
                top_left_x = random.randint(0, w - block_width)
                mask[i, 0, top_left_y:top_left_y + block_height, top_left_x:top_left_x + block_width] = 1
        elif mask_index == 7:
            center_x = torch.randint(0, w, (1,)).item()
            center_y = torch.randint(0, h, (1,)).item()
            a = torch.randint(min(w, h) // 8, min(w, h) // 4, (1,)).item()  # 长半轴
            b = torch.randint(min(h, w) // 8, min(h, w) // 4, (1,)).item()  # 短半轴

            for i in range(h):
                for j in range(w):
                    if ((i - center_y) ** 2) / (b ** 2) + ((j - center_x) ** 2) / (a ** 2) < 1:
                        mask[:, :, i, j] = 1
        elif mask_index == 8:
            center_x = torch.randint(0, w, (1,)).item()
            center_y = torch.randint(0, h, (1,)).item()
            radius = torch.randint(min(h, w) // 8, min(h, w) // 4, (1,)).item()
            for i in range(h):
                for j in range(w):
                    if (i - center_y) ** 2 + (j - center_x) ** 2 < radius ** 2:
                        mask[:, :, i, j] = 1
        elif mask_index == 9:
            for idx in range(f):
                if np.random.rand() > 0.5:
                    mask[idx, :, :, :] = 1
        else:
            raise ValueError(f"The mask_index {mask_index} is not define")
    else:
        if f != 1:
            mask[1:, :, :, :] = 1
        else:
            mask[:, :, :, :] = 1
    return mask
```

**Comment by luchaoqi** [comment]
@bubbliiiing I'm still a bit confused about the model's capabilities for the regular inpainting task — specifically, whether it can accept a mask video rather than just performing first-and-last-frame generation, as the task @juntaosun mentioned. From the inference code [here](https://github.com/aigc-apps/VideoX-Fun/blob/6183d41df8fbeb44f24b6c617efb8ef0ab690052/videox_fun/utils/utils.py#L270), it seems that the model (whether it's WAN or COG) is not designed to receive a mask video as input. 

Is that correct? Or is it theoretically possible to feed a mask video into the model — but the performance would be suboptimal?

**Comment by sudocoder123** [comment]
@bubbliiiing I have a similar question.

The Wan2.2-VACE-Fun-A14B model can be used for video inpainting using the examples/wan2.2_vace_fun/predict_v2v_mask.py script.

Can the Wan2.2-Fun-5B-InP or Wan2.2-Fun-A14B-InP models be used for video inpainting? Can a similar python script be written to do that?

Thank you.
